### PR TITLE
Initial memory profiler view.

### PR DESCRIPTION
**Pull request by terrylucas**
Initial cut at memory profiler.  More work to do and UX review as well.

image

### CODE REVIEW

File: packages/devtools/lib/src/memory/memory.dart [hunk 1]
```diff
@@ -1,14 +1,15 @@
-// Copyright 2018 The Chromium Authors. All rights reserved.
+// Copyright 2019 The Chromium Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
 import 'dart:async';
 import 'dart:collection';
 import 'dart:math' as math;
 
+import 'package:devtools/src/ui/icons.dart';
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory.dart]: you made the same mistake I've made 50X times. Here and elsewhere search and remove any
package:devtools imports replacing with relative imports.

File: packages/devtools/lib/src/memory/memory.dart [hunk 2]
```diff
@@ -144,6 +206,22 @@ class MemoryScreen extends Screen {
     }
   }
 
+  // TODO(terry): Need to record the GC event in the timeline.
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory.dart]: do we have to record it? I would hope we would still get a GC event back even if we manually triggered the GC. That way the wacky case of multiple memory profiler users would still work.

File: packages/devtools/lib/src/memory/memory.dart [hunk 3]
```diff
@@ -144,6 +206,22 @@ class MemoryScreen extends Screen {
...
+  Future<Null> _gcNow() async {
+    gcNowButton.disabled = true;
+
+    try {
+      // 'reset': true to reset the object allocation accumulators
+      await serviceManager.service.callMethod('_getAllocationProfile',
+          isolateId: _isolateId, args: {'gc': 'full'});
+    } catch (e) {
+      // TODO(terry): Need something probably logging?
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory.dart]: is this a message we want to display to users. There is an API in this code to display a toast for users with an error.
If not, the TODO and print is fine.

File: packages/devtools/lib/src/memory/memory.dart [hunk 4]
```diff
@@ -204,7 +282,10 @@ class MemoryScreen extends Screen {
   CoreElement createLiveChartArea() {
     final CoreElement container = div(c: 'section perf-chart table-border')
       ..layoutVertical();
-    memoryChart = MemoryChart(container);
+    memoryChart = MemoryChart(this);
+
+    container.add(memoryChart);
+
     memoryChart.disabled = true;
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory.dart]: style nit: move the container.add call to after the memory chart is set to be disabled.

File: packages/devtools/lib/src/memory/memory.dart [hunk 5]
```diff
@@ -327,71 +408,55 @@ class MemoryColumnSimple<T> extends Column<T> {
...
+      plotlyChart = MemoryPlotly(memoryGraph, this)..plotMemory();
+      chartCreated = true;
+    }
 
-    final int width = MemoryTracker.kMaxGraphTime.inMilliseconds;
-    final int right = data.samples.last.time;
+    while (data.samples.isNotEmpty) {
+      final HeapSample newSample = data.samples.removeAt(0);
+      plotlyChart.plotMemoryDataList([newSample.timestamp], [newSample.rss],
+          [newSample.capacity], [newSample.used], [newSample.external]);
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory.dart]: add a trailing comma to make dart style do a better job.

File: packages/devtools/lib/src/memory/memory.dart [hunk 6]
```diff
@@ -327,71 +408,55 @@ class MemoryColumnSimple<T> extends Column<T> {
...
-    final int top = (data.maxHeapData ~/ tenMB) * tenMB + tenMB;
+  void updateFrom(MemoryTracker data) {
+    if (!chartCreated) {
+      plotlyChart = MemoryPlotly(memoryGraph, this)..plotMemory();
+      chartCreated = true;
+    }
 
-    final int width = MemoryTracker.kMaxGraphTime.inMilliseconds;
-    final int right = data.samples.last.time;
+    while (data.samples.isNotEmpty) {
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory.dart]: nit: why isn't this a 
for (HeapSample sample in data.samples) {
...
}
data.samples.clear();

File: packages/devtools/lib/src/memory/memory_plotly.dart [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        symbol: 'circle',
+        size: 10,
+      ),
+      name: 'GC',
+      text: [],
+      hoverinfo: 'y+name',
+    );
+    */
+
+    final Data externalTrace = Data(
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory_plotly.dart]: add a helper method so it is easier to see what is different about each one of these Data objects. Most look the same except for the group, name, and color.

File: packages/devtools/lib/src/memory/memory_plotly.dart [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    List<int> timestamps,
+    List<num> rsses,
+    List<num> capacities,
+    List<num> uses,
+    List<num> externals,
+  ) {
+    // TODO(terry): Eliminate this JS call (result of reified List?).
+    myExtendTraces(
+      _domName,
+      timestamps,
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory_plotly.dart]: comment why timestamps is repeated 4 times.

File: packages/devtools/lib/src/memory/memory_plotly.dart [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+          tickformat: '%-I:%M:%S %p',
+          hoverformat: '%-I:%M:%S.%L %p',
+        ),
+      ),
+    );
+  }
+
+  bool liveUpdate = true;
+
+  void setLiveUpdate({bool live}) {
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory_plotly.dart]: this method doesn't seem like it adds any value. 
Just set
liveUpdate = live directly.

**Comment by devoncarew** [comment]
As some drive-by comments on the UI:

- I'd just use one 'primary' button on a page; you have two, I'd either choose the most semantically important button on the page, or choose a button that's also primary on another page
- you should probably have text for the buttons on the upper right, for similarity with the other pages
- we should ensure that the chart height is the same as the timeline page, so there's no big jump when  switching between the pages 
- buttons that are similar can be in button groups; button groups themselves (or individual buttons) should have a bit of padding between them

**Comment by devoncarew** [comment]
Perhaps swap the location of the memory and debugger pages? The Timeline view seems most similar to the memory one, so perhaps should be adjacent.

File: packages/devtools/lib/src/memory/memory.dart (hunk 1 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory.dart]: Will do I think it was automatically added by IntelliJ.

File: packages/devtools/lib/src/memory/memory.dart (hunk 2 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory.dart]: No, the GC itself will cause the event to fire.

File: packages/devtools/lib/src/memory/memory.dart (hunk 3 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory.dart]: I think just a toast that the GC couldn't be done is enough.  I'll print out the error detail to the console.

File: packages/devtools/lib/src/memory/memory.dart (hunk 4 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory.dart]: memoryChart = MemoryChart(this)..disabled = true;
container.add(memoryChart);

done.

File: packages/devtools/lib/src/memory/memory.dart (hunk 5 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory.dart]: done.

File: packages/devtools/lib/src/memory/memory.dart (hunk 6 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory.dart]: Should be.

File: packages/devtools/lib/src/memory/memory.dart (hunk 5 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory.dart]: Done.

File: packages/devtools/lib/src/memory/memory_plotly.dart (hunk 7 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory_plotly.dart]: Done.

File: packages/devtools/lib/src/memory/memory_plotly.dart (hunk 8 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory_plotly.dart]: x coords for each trace that corresponds to the y coords.  Added comment.

File: packages/devtools/lib/src/memory/memory_plotly.dart (hunk 9 above)
Comment (terrylucas) [reply packages/devtools/lib/src/memory/memory_plotly.dart]: Done.

**Comment by terrylucas** [comment]
> Perhaps swap the location of the memory and debugger pages? The Timeline view seems most similar to the memory one, so perhaps should be adjacent.

Done.

**Comment by terrylucas** [comment]
> Some drive-by comments on the UI:

Agree, changed to fix these suggestions.

File: packages/devtools/lib/src/memory/memory.dart (hunk 3 above)
Comment (jacob314) [reply packages/devtools/lib/src/memory/memory.dart]: we have support to display toasts. I would go ahead and display one.

File: packages/devtools/lib/src/memory/memory.dart [hunk 10]
```diff
@@ -35,6 +35,8 @@ class MemoryScreen extends Screen with SetStateMixin {
     addStatusItem(objectCountStatus);
   }
 
+  MemoryController memoryController = MemoryController();
```
Comment (jacob314) [inline packages/devtools/lib/src/memory/memory.dart]: nit: make final

**Pull request merged**